### PR TITLE
Pull and Push of Line and Node Constraints

### LINKED ISSUE #19: Pushing of Arc centre and midpoint nodes
#### Description:

When Pushing arc geometry into RFEM the centre and midpoints are pushed along with the nodes for the curve. These nodes are not required within RFEM for the lines. Having these nodes creates larger models, and more complicated post-processing of the model due to the amount of nodes that are present but not used. These nodes should be deleted after the push of arc elements

#### Steps to reproduce:

Push arc geometry in a bar or surface boundary definition. The nodes will then appear in the model.

#### Expected behaviour:

Arced geometry should be pushed without the unused nodes.

### LINKED ISSUE #24: Push/Pull of Contraints Node and Line Constraints
#### Description:

Structural model usually so need numberous Constraints. It would be beneficial to add funcionalities that do enable the user to push differnt kids of supports. This does mainly include the following Supports: 
1. Node
2. Line

### LINKED ISSUE #25: Add Bar Relsease
#### Description:

When it comes to defining connection between Bars BHoM does offer an attribute Release when defining Bars. A Similare concept has also been defined in RFEM6. This concept should also be inplementd into the RFEM6_Toolkit

### LINKED ISSUE #26: Push of uncecessary nodes.
#### Description:

When it comes to pushing curves int RFEM6 that are not simple lines but rather Arches or circles some additional Nodes get pushed into RFEM6. These nodes are not really necessary in RFEM6 as the data is structured differntly here. However, those additionaly pushed nodes could potatntially cause confustiona and and are unfenificial when it comes to userfriendlyness and model quality. 

#### Steps to reproduce:

Push panel defined by polycure outline that has minimun one arc in it. 

#### Expected behaviour:

Only points that are necessary for defining the outline of the panel should be visible in REFM6.

### PR DESCRIPTION
**Pull request by Arne-Martensen**
### Issues addressed by this PR

After compiling, the RFEM6_Toolkit should be able to generate NodalSupports and LineSupports in RFEM. BHoM doesn't differentiate between Nodal- and LineSupport, so pushing these into RFEM6 always requires either a Node with a corresponding support or an Edge with a corresponding support. All supports are of type Constraint6DOF. It's important to note that handling the push of a Constraint6DOF of type FullReleaseConstraint6DOF is not yet implemented.

Additionally this PR aims to allow users to Push and Pull Bar Releases. The Push and Pull actions for Bar Releases can only be executed in conjunction with an actual bar that is being pushed or pulled.

Initially, when pushing arches or circles, unwanted nodes representing the arc or circle's center were also pushed. This issue has been resolved with this pull request. Now, only the necessary nodes will be pushed into RFEM6. 

Closes #24 
Closes #26 
Closes #19
Closes #25 

### Test files

[Test File](https://burohappold-my.sharepoint.com/:u:/p/arne_martensen/EeZWVm0U-3JEnT2oGQaWsdEBsX8tTn6cCQiyAD20GBmISg?e=ocHT8A)

### Changelog

**For RFEMNodalSupport**
- Add Intermediate data structure RFEMNodalSupport
- Read Method for RFEMNodalSupport + Converter FromRFEM()
- Create Method for RFEMNodalSupport + Converter ToRFEM()
- Comparar for RFEMNodalSupport
- DependencyModule for RFEMNodalSupports: For initializing Objects of type RFEMNodalSupport

**For RFEMLineSupport**
- Add Intermediate data structure RFEMLineSupport
- Read Method for RFEMLineSupport+ Converter FromRFEM()
- Create Method for RFEMLineSupport+ Converter ToRFEM()
- Comparar for RFEMLineSupport
- DependencyModule for RFEMLineSupport: For initializing Objects of type RFEMLineSupport

**For RFEMHinge**
- Add Intermediate data structure RFEMHinge
- Read Method for RFEMHinge+ Converter FromRFEM()
- Create Method for RFEMHinge+ Converter ToRFEM()
- Comparar for RFEMHinge
- DependencyModule for RFEMHinge: For initializing Objects of type RFEMHinge

**Update RFEMWebServiceLibrary**
- The RFEMWebServiceLibrary version has been updated to version 6.3.7  

**Adding Base for Unit Test**
- Added an extra .sln file for adding potential unit tests

### Additional comments

- When pushing panels, it's recommended to attach the Constraint6DOF directly to the panel's defining edge. If you push the panel and its edges separately, you may run into conflicts. This is because panels inherently reference edges, not curves. If done incorrectly, this could cause conflicts between the panel edges and the edges affecting the line support. Additionally, it's recommended to use closed lines to define panels, as this is more likely to avoid contradictions between multiple defined panels or edges.

### CODE REVIEW

**Review by staintono — COMMENTED** [review]
Using the test script provided I tried to push nodal supports. This did not work and produced the following error: 
1. This component failed to run properly.
- Error: Object reference not set to an instance of an object.
- Occured in  BH.Adapter.RFEM6.RFEM6Adapter.Disconnect()
     called from BH.Adapter.RFEM6.RFEM6Adapter.Push(IEnumerable`1 objects, String tag, PushType pushType, ActionConfig actionConfig)
- Are you sure you have the correct type of inputs? Check their description for more details.

The code was not reviewed.

**Review by staintono — CHANGES_REQUESTED** [review]
Tested by creating a rectangular panel and extract its edges to use as line supports. The surface and the line support types are pushed correctly, however the lines/panel edges do not have the line support types associated with them.

**Comment by Arne-Martensen** [comment]
> Tested by creating a rectangular panel and extract its edges to use as line supports. The surface and the line support types are pushed correctly, however the lines/panel edges do not have the line support types associated with them. Using the same process but with nodes, the node supports are correctly pushed and assigned.

In your script, I noticed you're trying to push the same edge twice: once as part of the panel and once separately, but with constraints. When creating a panel using outlines (not edges), it's important to know that the panel component internally defines edges based on your input curve. So, to the computer, it appears you're pushing the same edge twice: once with constraints and once without. This causes a conflict, and the computer has to choose one. In your case, it seems the edge for support was ignored when creating the geometry, but used only to define LineSupport. Meanwhile, the panel edge defined the geometry. This conflict results in a LineSupport that isn't aware of its connection to the panel.

I'll have another look but it is important to point out that this approach may increase computational load, as it requires looping and comparing edges for equality. 

A quick solution to your issue is to define the panel using edges that already have the constraints attached.

**Review by staintono — DISMISSED** [review]
The PR was reviewed using the provided test scripts as well as some user created geometries. The following elements were tested successfully with both pushing and pulling: Nodal and Line Supports with straight lines, polylines and arcs, Bar Releases. Line supports that coincide with Surface outlines could only be pushed when using edges to define the surface. Arcs are now pushed without the helper nodes and now only contain the minimum needed to define the arc. 
When pulling supports with spring stiffnesses were pulled as Free supports rather than fixed, however due to the fact that the toolkit currently does not support spring stiffnesses this will be raised as a separate issue. 
The Code was not reviewed.

**Review by staintono — DISMISSED** [review]
The PR was tested using the provided test scripts as well as user created models. The following elements were tested and can be successfully pushed and pulled: Nodal Supports, Line Supports, Bar Hinges. The Line supports can only be pushed when pushed as part of an Edge that defines a panel. Arcs can now be pushed without the helper nodes being present after the push. The code was not checked

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance
@BHoMBot check required

File: RFEM6_Adapter/CRUD/Create/Node.cs [hunk 1]
```diff
@@ -42,25 +42,45 @@ private bool CreateCollection(IEnumerable<Node> bhNodes)
...
+        }
+
+        private static DOFType tranlate(double input)
+        {
+
+            if (input.Equals(double.PositiveInfinity)) return DOFType.Fixed;
+            else return DOFType.Free;
+
+
+
```
Comment (FraserGreenroyd) [inline RFEM6_Adapter/CRUD/Create/Node.cs:82]:
```suggestion
            if (input.Equals(double.PositiveInfinity))
                return DOFType.Fixed;
            else
                return DOFType.Free;
```

File: RFEM6_Adapter/CRUD/Create/Node.cs [hunk 2]
```diff
@@ -42,25 +42,45 @@ private bool CreateCollection(IEnumerable<Node> bhNodes)
...
-                m_Model.set_node(rfNode);
             }
             return true;
         }
 
 
+        private static bool ComparerRFEMSupportAndBHoMConstraint(rfModel.nodal_support rfSupport, Constraint6DOF bhConstraint)
+        {
+
+            if (!tranlate(rfSupport.spring_x).Equals(bhConstraint.TranslationX)) { return false; }
```
Comment (FraserGreenroyd) [inline RFEM6_Adapter/CRUD/Create/Node.cs:65]: Please put all commands from an `if` onto its own line - no need for brackets if only a single line return is used.

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Please resolve the compliance issues highlighted by the bot as well.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check serialisation
@BHoMBot check core
@BHoMBot check installer
@BHoMBot check null-handling

**Review by FraserGreenroyd — APPROVED** [review]
Approving based on [this review](https://github.com/BHoM/RFEM6_Toolkit/pull/29#pullrequestreview-1663984578) as remaining commits have been for compliance and rebasing.

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am authorising dispensation to be granted on check ref. 17520982879

**Pull request merged**